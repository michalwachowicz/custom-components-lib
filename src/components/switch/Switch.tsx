import { useId, useRef } from "react";
import * as styles from "./Switch.module.scss";
import Switchable from "../../interfaces/Switchable";
import useClickCenterEntity from "../../hooks/useClickCenterEntity";
import Entity from "../clickEntity/ClickEntity";

export default function Switch({
  ref,
  label = undefined,
  checked = false,
  disabled = false,
  onChange,
}: Switchable) {
  const id = useId();
  const handleRef = useRef<HTMLDivElement>(null);
  const { entity, size, spawnEntity } = useClickCenterEntity(handleRef);

  const handleChange = () => {
    if (disabled) return;
    if (onChange) onChange(!checked);

    spawnEntity();
  };

  const checkedClass = checked ? styles["switch-checked"] : "";
  const disabledClass = disabled ? styles["switch-disabled"] : "";

  return (
    <div className={styles["switch-container"]}>
      <input
        ref={ref}
        id={id}
        className={styles["switch-checkbox"]}
        type='checkbox'
        checked={checked}
        disabled={disabled}
        onChange={handleChange}
      />

      <div
        data-testid='switch'
        aria-hidden='true'
        className={`${styles.switch} ${checkedClass} ${disabledClass}`}
        onClick={handleChange}
      >
        <div className={styles["switch-back"]}></div>
        <div ref={handleRef} className={styles["switch-handle"]}>
          <div className={styles["switch-handle-inner"]} />

          {entity !== null && (
            <Entity key={entity.id} size={size} entity={entity} />
          )}
        </div>
      </div>

      {label && (
        <label className={styles["switch-label"]} htmlFor={id}>
          {label}
        </label>
      )}
    </div>
  );
}
