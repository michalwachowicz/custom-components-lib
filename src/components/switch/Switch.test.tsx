import { act } from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Switch from "./Switch";

describe("<Switch />", () => {
  it("renders with label properly", () => {
    render(<Switch label='Test' />);

    expect(screen.getByText("Test")).toBeInTheDocument();
  });

  it("is checked when checked prop is true", () => {
    render(<Switch checked />);
    expect(screen.getByRole("checkbox")).toBeChecked();
  });

  it("is not checked when checked prop is false", () => {
    render(<Switch />);
    expect(screen.getByRole("checkbox")).not.toBeChecked();
  });

  it("is disabled when disabled prop is true", () => {
    render(<Switch disabled />);
    expect(screen.getByRole("checkbox")).toBeDisabled();
  });

  it("is not disabled when disabled prop is false", () => {
    render(<Switch />);
    expect(screen.getByRole("checkbox")).not.toBeDisabled();
  });

  it("runs onChange when clicked", async () => {
    const onChange = jest.fn();
    const user = userEvent.setup();

    render(<Switch onChange={onChange} />);
    expect(onChange).toHaveBeenCalledTimes(0);

    await act(async () => {
      await user.click(screen.getByTestId("switch"));
    });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(true);
  });

  it("does not run onChange when disabled", async () => {
    const onChange = jest.fn();
    const user = userEvent.setup();

    render(<Switch onChange={onChange} disabled />);

    await act(async () => {
      await user.click(screen.getByRole("checkbox"));
    });

    expect(onChange).toHaveBeenCalledTimes(0);
  });

  it("spawns entity on click", async () => {
    jest.useFakeTimers();

    render(<Switch />);
    const user = userEvent.setup({ delay: null });

    await act(async () => {
      await user.click(screen.getByTestId("switch"));
    });

    expect(screen.getByTestId("entity")).toBeInTheDocument();

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.queryByTestId("entity")).not.toBeInTheDocument();
    jest.useRealTimers();
  });

  it("does not spawn entity when disabled", async () => {
    render(<Switch disabled />);

    const user = userEvent.setup();

    await act(async () => {
      await user.click(screen.getByTestId("switch"));
    });

    expect(screen.queryByTestId("entity")).not.toBeInTheDocument();
  });

  it("toggles the switch with the spacebar", async () => {
    const user = userEvent.setup();
    const handleChange = jest.fn();

    render(<Switch onChange={handleChange} />);
    expect(handleChange).toHaveBeenCalledTimes(0);

    const checkbox = screen.getByRole("checkbox");

    await act(async () => {
      await user.tab();
    });

    expect(checkbox).toHaveFocus();

    await act(async () => {
      await user.keyboard("[Space]");
    });

    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  describe("snapshots", () => {
    const labels = [undefined, "Test"];
    const booleans = [true, false];

    const testCases = labels.flatMap((label) =>
      booleans.flatMap((checked) =>
        booleans.map((disabled) => ({
          label,
          checked,
          disabled,
        }))
      )
    );

    test.each(testCases)(
      "renders correctly with label=%s, checked=%s, disabled=%s",
      ({ label, checked, disabled }) => {
        const { asFragment } = render(
          <Switch label={label} checked={checked} disabled={disabled} />
        );

        expect(asFragment()).toMatchSnapshot();
      }
    );
  });
});
