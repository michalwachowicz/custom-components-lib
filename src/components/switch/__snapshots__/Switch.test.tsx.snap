// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> snapshots renders correctly with label={ label: 'Test', checked: false, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      class="switch-checkbox"
      id="«rh»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch  "
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
    <label
      class="switch-label"
      for="«rh»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: 'Test', checked: false, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      class="switch-checkbox"
      disabled=""
      id="«rg»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch  switch-disabled"
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
    <label
      class="switch-label"
      for="«rg»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: 'Test', checked: true, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      checked=""
      class="switch-checkbox"
      id="«rf»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch switch-checked "
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
    <label
      class="switch-label"
      for="«rf»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: 'Test', checked: true, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      checked=""
      class="switch-checkbox"
      disabled=""
      id="«re»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch switch-checked switch-disabled"
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
    <label
      class="switch-label"
      for="«re»"
    >
      Test
    </label>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: undefined, checked: false, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      class="switch-checkbox"
      id="«rd»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch  "
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: undefined, checked: false, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      class="switch-checkbox"
      disabled=""
      id="«rc»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch  switch-disabled"
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: undefined, checked: true, disabled: false }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      checked=""
      class="switch-checkbox"
      id="«rb»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch switch-checked "
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Switch /> snapshots renders correctly with label={ label: undefined, checked: true, disabled: true }, checked=%s, disabled=%s 1`] = `
<DocumentFragment>
  <div
    class="switch-container"
  >
    <input
      checked=""
      class="switch-checkbox"
      disabled=""
      id="«ra»"
      type="checkbox"
    />
    <div
      aria-hidden="true"
      class="switch switch-checked switch-disabled"
      data-testid="switch"
    >
      <div
        class="switch-back"
      />
      <div
        class="switch-handle"
      >
        <div
          class="switch-handle-inner"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
